DataViews: Add user input filter

chihsuan opened the pull request:
<!--
Link a related Github/Linear issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword. Consider using "part of" instead.
-->

Part of WOOA7S-123

## Proposed Changes


Add support for user input in filters


## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

To support dynamic filtering for integer fields and other types that benefit from user input, aligning with ongoing improvements to DataViews filtering capabilities.

Some use cases require filters that accept user-provided input — like "Price greater than 100", or "Created before 2023-01-01". Right now, the package doesn’t allow this, making it hard to support numeric or textual search criteria.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

*  `yarn dataviews:storybook:start`
* Go to `http://localhost:57863/?path=/story/dataviews-dataviews--default`
* Open the filter panel and select different fields to see if the filter works as expected.



## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

chihsuan committed:
Add userInput property to FilterByConfig and NormalizedFilter interfaces

This update introduces a new optional property, userInput, to both the FilterByConfig and NormalizedFilter interfaces, allowing filters to accept user input. Additionally, a type property is added to NormalizedFilter for better field type specification.

chihsuan committed:
Add UserInputWidget component for filter input handling

This commit introduces the UserInputWidget component, which allows users to input values for filters in the dataviews. It includes a SingleInputSummary for managing single selection operators and utilizes a debounced input for better performance. The component integrates with existing filter logic to update the view's filters based on user input.

chihsuan committed:
Enhance FilterSummary to support UserInputWidget for dynamic filtering

This update modifies the FilterSummary component to conditionally render the UserInputWidget based on the userInput property of filters. It also adjusts the logic for determining active elements to accommodate user input, ensuring a seamless integration with existing filter functionalities. Additionally, the useFilters hook is updated to handle the new userInput property, and styles are added for the UserInputWidget.

chihsuan committed:
Refine operator filtering logic in sanitizeOperators function

This update enhances the sanitizeOperators function to specifically filter operators for integer fields. Additionally, the logic in the column header menu is adjusted to allow adding filters based on user input or existing elements, improving the overall filtering capabilities in the dataviews.

chihsuan committed:
Add filterBy configuration for integer fields to support user input

This update introduces a filterBy property for integer fields in the fixtures, allowing the use of specific operators and enabling user input for filtering. This enhancement aligns with previous updates to improve filtering capabilities in the dataviews.

chihsuan committed:
Refactor UserInputWidget to streamline filter handling

This update refines the UserInputWidget component by renaming the currentFilter to activeFilter for clarity and consistency. The applyFilterChange function is simplified to directly handle the new value logic, enhancing readability. Additionally, the SingleInputSummary component is updated to ensure it only renders when an active filter is present and valid, improving the overall user experience in managing filters.

youknowriad commented on packages/dataviews/src/types.ts:
```diff
@@ -232,11 +237,21 @@ export interface NormalizedFilter {
+	/**
+	 * Whether the filter allows user input instead of selecting from the list of options.
+	 */
+	userInput?: boolean;
```
Why is this needed? Can we just fallback to the `type`'s `Edit` function (which is already a user input) when the "type" is defined.

chihsuan replied:
We introduced `userInput` prop because we want to differentiate the user input from the preset options. Previously, the filter [can only be added if field has element](https://github.com/Automattic/wp-calypso/pull/103276/files#diff-7dfdcfa8718e6c4e851072624678f77d04463388b6b387186657cc32d6e8c7b6L91). But for integer filer, users should be able to enable filter without adding any elements. 

Enabling the filter based on the field type is an option, but it might not provide enough flexibility. For the integer type, I think these scenarios should be supported:

1. No filter needed for the field
2. Filter with preset options (set elements)
3. Filter with user input (userInput: true)
4. Filter with preset options and user input (not sure if this is needed?)

Does this make sense? Or we should always enable the filter for integer type?

youknowriad replied:
For me if there's no elements, it should just fallback automatically to userInput, so the prop might not be needed at all.

oandregal replied:
I've left some comment below on the API that is [worth checking](https://github.com/Automattic/wp-calypso/pull/103276#discussion_r2084875336), but iterating on it and addressing the states we want to support:

> 1. No filter needed for the field.

```js
{
  id: 'id', 
  type: 'integer'
}
```

> 2. Filter with preset options and user input.

This is the default experience for typed fields that provide `elements`. It's the same for `date`.

```js
{
  id: 'id', 
  type: 'integer',
  elements: [ ... ]
}
```

> 3. Filter with only user input.

```js
{
  id: 'id',
  type: 'integer',
  filterBy: {
    input: true
  }
}
```

> 4. Filter with only presets

```js
{
  id: 'id',
  type: 'integer',
  elements: [ ... ],
  filterBy: {
    input: false
  }
}
```

chihsuan replied:
Thanks for the feedback! @youknowriad @oandregal 

For this PR, I'll go with the approach that @oandregal [suggested](https://github.com/Automattic/wp-calypso/pull/103276#discussion_r2084875336).

> - if field.type is integer, display the Edit control for the filter (input control for integers, by default, but the consumer can provide its own).
> - if field.elements is provided, display presets as well.

For Woo Analytics, the design requires support for hiding the input control. However, I'll double check with the designer. If the designer still prefers hiding the input control, I'll work on it in a separate PR.

youknowriad replied:
Not sure I like the "input" config personally, I'd rather start opinionated: by default "input", if there are elements "presets", no mix of both, and potentially have a way to disable a field from being a filter.

chihsuan replied:
> I'd rather start opinionated: by default "input", if there are elements "presets", no mix of both, and potentially have a way to disable a field from being a filter.

This sounds good to me. We don't have use cases for both "input" and "presets" at the same time. 

Quick summary:

| Field Configuration                                | `filterBy.enabled` | `elements` Present | Rendered Filter UI |
|----------------------------------------------------|---------------------|---------------------|---------------------|
| `{ type: 'integer' }`                              | not set (default)   | ❌                  | Input               |
| `{ type: 'integer', elements: [...] }`             | not set (default)   | ✅                  | Presets             |
| `{ type: 'integer', filterBy: { enabled: false }}` | false               | ❌ or ✅             | No filter           |

oandregal replied:
> We don't have use cases for both "input" and "presets" at the same time.

Yes, we do :) The date filter is one use case, for example. This is the default experience in my view.

> { type: 'integer' }

I don't think this field configuration should render any filter. Filter have been "opt-in" so far, and I think it makes more sense for them to stay that way. To opt-in, either you provide elements or you enable something so it displays the Edit/input. Additionally, consumers should be able to disable the input/Edit or the presets.

Based on the use cases you mentioned before, I'll elaborate a bit more on how I see this working:

- By default, a field doesn't provide any filter:

```js
{ type: 'integer' } 
```

- If it has elements, it provides a filter with input/Edit and presets (this is the default experience):

```js
{ type: 'integer', elements: [ ... ] }
```

- It should be possible to disable the input or the presets:

```js
{ type: 'integer', elements: [ ... ], filterBy: { widget: [ 'presets' ] } }  // only enable presets
{ type: 'integer', elements: [ ... ], filterBy: { widget: [ 'edit' ] } }  // only enable edit
```

- It should be possible to enable the filter even if the field doesn't have elements:

```js
{ type: 'integer', filterBy: { widget: [ 'edit' ] } } 
```

I care about supporting this experiences, but I don't mind much about using `filterBy.input`, `filterBy.widget` or any other naming. I'll be AKF next week, so I won't be able to provide feedback, but wanted to share how I expect this to work.

chihsuan replied:
> Yes, we do :) The date filter is one use case, for example. This is the default experience in my view.

Ah, you're right! The date filter does require both input and presets. 🤔 

> I care about supporting this experiences, but I don't mind much about using filterBy.input, filterBy.widget or any other naming. 

I don't mind the naming either. I think the key point is whether the filter should be enabled or disabled by default.

Since we want to support input and presets, I'd agree to disable the filter by default to avoid clutter and confusion; no filter is rendered unless the consumer explicitly opts in.

youknowriad reviewed (commented):
Thanks for working on this, the behavior seems correct but the abstraction/implementation not yet. Check the other comments and let me know what you think.

oandregal commented on packages/dataviews/src/components/dataviews-filters/filter-summary.tsx:
```diff
@@ -329,7 +342,11 @@ export default function FilterSummary( {
 					<VStack spacing={ 0 } justify="flex-start">
 						<OperatorSelector { ...commonProps } />
-						<SearchWidget { ...commonProps } />
+						{ filter.userInput ? (
```
Why is this checking for a new prop `userInput`? Can we use the field type instead? For context, that's what I've explored in the [date filter](https://github.com/Automattic/wp-calypso/pull/103136/files#diff-66986a72a5a66aa4d65ca24ca5dc95d4b836836aa86630a10b9685b7a2fe12c5R449).

chihsuan replied:
We use `userInput` prop because if users only provide preset options, we should render SearchWidget. We can't use the field type to determine if we should render UserInputWidget or SearchWidget.


https://github.com/Automattic/wp-calypso/pull/103276#discussion_r2084019792

oandregal replied:
> We use userInput prop because if users only provide preset options, we should render SearchWidget

This is how I'd expect it to work:

- if `field.type` is `integer`, display the `Edit` control for the filter (input control for integers, by default, but the consumer can provide its own).
- if `field.elements` is provided, display presets as well.

Do you need support for the case of hiding the input control? Elements without input is how the current filters work (untyped fields). If you want to add that in this PR, I'd suggest a generic approach that works for any field type later. For example, adding a new flag under filterBy to only render presets (the default experience is using everything):

```js
{
  id: '...',
  type: 'integer',
  elements: [ ... ]
  filterBy: {
    presetsOnly: true
  }
}
```

Additionally, I think the `SearchWidget` should just have different widgets per field type:

```js
if ( props.filter.type === 'datetime' ) { ... }
if ( props.filter.type === 'integer' ) { ... }

// Default widgets that just render elements.
```

As we iterate on this, using this approach would be clearer and has the following advantage: when/if we allow 3rd parties to define their own field types, they'll just work if we're able to make the filter part of the field definition. It's good to have it in mind, but it's not a requirement to implement it like that right now.

oandregal replied:
Continued the conversation above https://github.com/Automattic/wp-calypso/pull/103276#discussion_r2084902023

oandregal commented on packages/dataviews/src/components/dataviews/stories/fixtures.tsx:
```diff
@@ -682,6 +682,10 @@ export const fields: Field< SpaceObject >[] = [
 		type: 'integer',
 		enableSorting: true,
+		filterBy: {
+			operators: [ 'is', 'isNot' ],
```
This is another thing I've run into with the date filter. Consumers shouldn't need to declare this just because an `integer` doesn't support `isAny` or `isNone`, it should be the field type that provides a list of "supported operators". As a first step, I've added the logic to the `sanitizeOperators` in the date filter PR, but I think it could be part of the filed type definition itself.

oandregal replied:
Another question is: what does it mean that a field of type `integer` provides a list of `elements`? Do we ignore them? Or is it a way to provide "presets"?

chihsuan replied:
> but I think it could be part of the filed type definition itself.

Sounds good. I'll explore the idea of adding a list of supported operators to the field type definition. 👍

> Another question is: what does it mean that a field of type integer provides a list of elements? Do we ignore them? Or is it a way to provide "presets"?

It's a way to provide "presets". We still allow consumers to declare a list of elements for a field of type integer, like a text field does.

chihsuan replied:
I've update the PR with the following changes:

- The `userInput` prop has been removed and the field type is now used to determine if the filter should be a user input.
- The operators logic has been moved from `sanitizeOperators` to the field type definition.
- The control is now rendered using the `Edit` field.

Could you please take a look and let me know what you think?

chihsuan committed:
Refactor filtering components to enhance user input handling

chihsuan committed:
Update SearchWidget to refine integer filter handling

This change modifies the SearchWidget component to ensure that the InputWidget is only rendered for integer filters when no elements are present. This adjustment enhances the filtering logic and improves user experience by preventing unnecessary input fields from appearing.

youknowriad commented on packages/dataviews/src/components/dataviews-filters/filter-summary.tsx:
```diff
@@ -235,12 +238,25 @@ export default function FilterSummary( {
+
+	if (
+		activeElements.length === 0 &&
+		filter.type === 'integer' &&
```
Why are we hard-coding the type here?

chihsuan replied:
See below. https://github.com/Automattic/wp-calypso/pull/103276#discussion_r2090350760

youknowriad commented on packages/dataviews/src/components/dataviews-filters/index.tsx:
```diff
@@ -28,7 +28,7 @@ export function useFilters( fields: NormalizedField< any >[], view: View ) {
 		const filters: NormalizedFilter[] = [];
 		fields.forEach( ( field ) => {
-			if ( ! field.elements?.length ) {
+			if ( ! field.elements?.length && field.type !== 'integer' ) {
```
Why are we hard coding the type here?

youknowriad replied:
It should be sufficient to check if the current field type as an "edit" function defined.

chihsuan replied:
Because this PR only adds support for the integer type, and from my understanding, we will iterate on this to add support for other types, as @oandregal [commented](https://github.com/Automattic/wp-calypso/pull/103276#discussion_r2084875336). He is also working on another PR to add support for datetime types.

youknowriad replied:
For me, this should test the presence of an "Edit" function in the "type" instead of actually hardcoding types.

youknowriad commented on packages/dataviews/src/components/dataviews-filters/search-widget.tsx:
```diff
@@ -319,6 +326,10 @@ function ComboboxList( { view, filter, onChangeView }: SearchWidgetProps ) {
 
 export default function SearchWidget( props: SearchWidgetProps ) {
+	if ( props.filter.type === 'integer' && ! props.filter.elements ) {
+		return <InputWidget { ...props } />;
```
I'm not sure I understand what's happening here and why we're creating this new "InputWidget" component, I believe we should just rely on the "edit" defined by the field type.

chihsuan replied:
The `InputWidget` component is created for managing the onChange event and the filter value. It does rely on the `Edit` component within it.

chihsuan replied:
Additionally, the `InputWidget` component will accommodate operators such as `between` and `not between` (we 
 plan to add these in another PR) for integer types, which require the rendering of two `Edit` components.

youknowriad replied:
Makes sense, but we shouldn't be hard coding the "type", we should check the presence of "edit" and the config of the "type" instead. It's important to understand that when we add a new "type" to dataviews, we shouldn't have to touch any other file in the library, registration of types should be decoupled from the framework itself.

chihsuan replied:
Okay, thanks for the clarification. 🙏 

I thought we wanted to add support for the integer field first, and gradually add support for other field types. The type logic is temporary and will be removed in the future. 🤔 

> Makes sense, but we shouldn't be hard coding the "type", we should check the presence of "edit" and the config of the "type" instead. 

Then, this PR will affect all field types. Any fields with an Edit Prop will enable filter support. cc @oandregal

chihsuan replied:
The NormalizeField appears to guarantee the presence of the `Edit` prop. In this case, we should verify that the Edit returns non-null elements, right?

https://github.com/Automattic/wp-calypso/blob/a68a7adb0392201600363d27b7b30c682a234787/packages/dataviews/src/types.ts#L188

youknowriad replied:
> The NormalizeField appears to guarantee the presence of the Edit prop. In this case, we should verify that the Edit returns non-null elements, right?

This is interesting, you're right, it does seem like it's always defined. Personally, I think we should change the normalized field to have a `null` Edit when there's no edit defined (basically when the field has no "type" or the field has no explicit "Edit" function).

> I thought we wanted to add support for the integer field first, and gradually add support for other field types. 

You're right, making this change will enable the direct input for all types, I think it's a good thing though.

oandregal replied:
> Makes sense, but we shouldn't be hard coding the "type", we should check the presence of "edit" and the config of the "type" instead. It's important to understand that when we add a new "type" to dataviews, we shouldn't have to touch any other file in the library, registration of types should be decoupled from the framework itself.

As I [mentioned](https://github.com/Automattic/wp-calypso/pull/103276#discussion_r2084875336), I agree with this direction, but I am fine if we start limiting by type for now (making it work based on the Edit requires a few big changes, so starting with the types helps contain scope). I'm also fine if we want to make it all work for all types at once in a single PR.

chihsuan replied:
Thanks, I've updated the PR as suggested.

- Changed `Edit` Prop type to accept `null`
- Added user input filter support based on the `Edit` property
- Added default filterBy operators for each field type
- Updated PR description and changelog to reflect the changes

https://github.com/user-attachments/assets/8617340e-622f-435f-91b1-9ae0002a61ff

This PR doesn't add a ability to disable the filter yet. I'll add that in a separate PR. Please review when you have time and let me know if anything else needs to be changed. 🙏

chihsuan committed:
Merge trunk

chihsuan committed:
Add changelog

chihsuan committed:
Enhance filter functionality by adding user input support based on the Edit

Update changelog to reflect this change.

chihsuan committed:
Refactor input and search widgets to utilize fields prop for improved functionality

chihsuan committed:
Fix type

chihsuan committed:
Merge branch 'trunk' into add/dataviews-filter-integer

chihsuan committed:
Merge trunk

oandregal commented:
In the current status, the PR generates filters for all fields that have a type or elements. This is problematic: consider that we are using this package in some places; merging this PR would activate unintended filters for calypso consumers whose fields declare a type (e.g., dotcom sites, agencies in multiple places, jetpack plugins).

Refer to [this comment](https://github.com/Automattic/wp-calypso/pull/103276#discussion_r2092601407) for how I'd expect it to work.

https://github.com/user-attachments/assets/33e34add-cec6-4006-865c-00e62bdaa7a4

oandregal commented on packages/dataviews/src/dataviews-layouts/table/column-header-menu.tsx:
```diff
@@ -84,11 +84,12 @@ const _HeaderMenu = forwardRef( function HeaderMenu< Item >(
 	canAddFilter =
 		! view.filters?.some( ( _filter ) => fieldId === _filter.field ) &&
-		!! field.elements?.length &&
+		!! ( field.elements?.length || field.Edit ) &&
```
There are existing filters in dotcom sites, agencies (multiple places) and jetpack cloud that use this package directly. We need to be opt-in.

We also need to engage design to confirm displaying the edit controls in the existing filters is ok (may require updating the Edit for existing types, etc.). Otherwise, we need to add the "disable edit" feature in this PR and also update the existing places to only use presets, first.

chihsuan replied:
> There are existing filters in dotcom sites, agencies (multiple places) and jetpack cloud that use this package directly. We need to be opt-in.

In that case, IMO we should disable the edit by default even if `elements` prop is provided to avoid any issues.  Consumer should explicitly opt-in to this edit feature.

youknowriad replied:
Can you explain more @oandregal what issues this creates?

chihsuan replied:
I believe @oandregal's concern is that enabling input filters by default, particularly when fields include Edit controls, might lead to problems for existing users of this package. 

For example, dotcom sites currently use only one filter enabled by the elements prop. Upgrading to the new version might show more filters. André suggests [disabling input by default](https://github.com/Automattic/wp-calypso/pull/103276#discussion_r2092601407) or adding a property to disable it in this PR.

chihsuan replied:
> We also need to engage design to confirm displaying the edit controls in the existing filters is ok (may require updating the Edit for existing types, etc.).


@Automattic/dotcom-design 👋 This PR adds user input filters by using `Edit` controls for rendering the filters UI. The goal is to provide a way to filter data by user input. By using `Edit` controls, we can provide a more flexible and powerful way for customers to filter data, such as Advanced DateTime filters. Does this look good to you?

Here is video demo of the input filters:

https://github.com/user-attachments/assets/8617340e-622f-435f-91b1-9ae0002a61ff

And we will use this feature to build the filters (figma: hdK9XLiOefugy2bxalKBN1-fi-2795_100658) for WooCommerce Analytics.

oandregal commented on packages/dataviews/src/components/dataviews-filters/input-widget.tsx:
```diff
@@ -0,0 +1,73 @@
+	return (
+		<div className="dataviews-filters__user-input-widget">
+			<field.Edit
+				data={ { [ field.id ]: currentValue } }
```
This Edit function doesn't have access to the whole object, just one value. This is not a restriction we have in DataForm, where the Edit is used. What if the Edit access values other than the `field.id` (e.g., "price" + "discount" to generate a "consolidated price")?

chihsuan replied:
I've changed how InputWidget passes data to Edit, providing all filter values instead in 5920eedd86b6b8e52ee98c4939332333fa45b165.

oandregal commented on packages/dataviews/src/types.ts:
```diff
@@ -240,7 +245,7 @@ export interface NormalizedFilter {
 	 * The list of options to pick from when using the field as a filter.
 	 */
-	elements: Option[];
+	elements?: Option[];
```
The idea of NormalizedFilter (or NormalizedField) is that all props should exist, so code that deals with this type doesn't need to do checks for existance (we're normalized them before). In this case, why does it need to change to undefined? I haven't checked all the code, but I suppose we're using this to check something. Can't we check for size of the array?

chihsuan replied:
I changed it to optional because the `useFilters` doesn't normalize the `elements` prop, it's pass through as is so it could be undefined or an empty array. If that's the case, then it's a bug. I'll revert the type and fix the bug to ensure the elements are always defined. We can check the size.

https://github.com/Automattic/wp-calypso/blob/b1fa25b44c054a32b0b5f1b417aa2a885eddad25/packages/dataviews/src/components/dataviews-filters/index.tsx#L44

chihsuan replied:
Fixed `filter.elements` type and checked array size instead. 58f3f5e88c5087a51366db86806b14ec042c20a3

oandregal commented on packages/dataviews/src/types.ts:
```diff
@@ -175,12 +180,12 @@ export type Field< Item > = {
 	getValue: ( args: { item: Item } ) => any;
 	render: ComponentType< DataViewRenderFieldProps< Item > >;
-	Edit: ComponentType< DataFormControlProps< Item > >;
+	Edit: ComponentType< DataFormControlProps< Item > > | null;
```
Same thought here about "NormalizedField".

Additionally, if the filter is opt-in, as [suggested](https://github.com/Automattic/wp-calypso/pull/103276#discussion_r2092601407), do we need to change `Edit`? We'd check for the presence of `elements` of a `filterBy` config instead of Edit, no?

chihsuan replied:
Modifying the Edit function could prevent situations where the filter is opt-in, yet the Edit function displays empty content. Certain type definitions within the Edit function can lead to this issue. I'll check the code and think deeper about this.

oandregal commented on packages/dataviews/src/types.ts:
```diff
@@ -70,13 +70,18 @@ export type FieldTypeDefinition< Item > = {
+	/**
+	 * The filter config for the field.
+	 */
+	filterBy?: FilterByConfig | undefined;
```
I consider FieldDefinition as "good defaults" for each type. Instead of allowing it to be undefined, can we making the `filterBy.operators` mandatory for every field type (and also for the untyped fallback)? It's okay if the fieldtype definition says "no operator" (empty set).

chihsuan replied:
Sure, I'll make the operators mandatory.

chihsuan replied:
Updated `filterBy` property to be required and set with default operators in field types in 09072f31e0b78c19bcb7ec18e66bb368619e7dfb.

oandregal commented:
Considering [this comment](https://github.com/Automattic/wp-calypso/pull/103276#discussion_r2092601407) as how we want it to work, this PR still needs to allow rendering both Edit and presets. We'd need to add a "custom" preset in that scenario that's selected when the user interacts with the Edit function, as in this example:

https://github.com/user-attachments/assets/59ea9ccb-c0b0-4e5d-aa60-fc62b7c99a6d

youknowriad commented on packages/dataviews/src/components/dataviews-filters/filter-summary.tsx:
```diff
@@ -229,19 +232,33 @@ function OperatorSelector( {
 export default function FilterSummary( {
```
I feel the name of this component is a bit weird, this seems to be the actual filter (not a summary)

chihsuan replied:
Indeed, it can be confusing. This is the code where the component is used. Renaming it to 'Filter' would be clearer. I'll rename it.

https://github.com/Automattic/wp-calypso/blob/e29a70a654a8d75e388860e171a73d2026b3c701/packages/dataviews/src/components/dataviews-filters/index.tsx#L195-L208

chihsuan replied:
Renamed FilterSummary component to Filter in fe6fadaf6494abbbee1f32eb706633c20f2f911c.

youknowriad reviewed (commented):
I didn't test everything, but this is looking good to me.

youknowriad commented:
When it comes to "presets" and "direct input". It doesn't seem like it's "configuration based". If you take a look at the "date" filter example, the presets are not shown in a select box and then we show the "edit" of the date type. It's a completely advanced UI.

In other words, the `Date` Edit function is received "elements" as a prop kind of, and rendering a custom input with presets on the side.

For the other types it's more "normalized", it's either "elements" in a select/listbox or a simple input.

So I'm actually wondering the solution is to remove the automatic handling of "elements" from the "framework" and move it to the "Edit" function/controls instead. 

Maybe it's fine to not address this immediately in this PR and think it properly with the Date "elements" support separately.

chihsuan committed:
Merge branch 'trunk' into add/dataviews-filter-integer

chihsuan committed:
Fix types

chihsuan committed:
Rename FilterSummary component to Filter

chihsuan committed:
Fix types

chihsuan committed:
Pass all filter data to Edit

chihsuan committed:
Revert filter.elements type and check length instead

chihsuan committed:
Update filterBy property to be required and initialize with default operators in field types

chihsuan commented:
> In other words, the Date Edit function receives "elements" as a prop kind of, and renders a custom input with presets on the side.
> For the other types it's more "normalized", it's either "elements" in a select/listbox or a simple input.

Yeah, we will also need to adjust the Edit function to show different layouts based on the operator.

<img width="346" alt="Screenshot 2025-05-20 at 13 25 45" src="https://github.com/user-attachments/assets/54de2661-1c11-41e1-92ff-bd0d90b3f013" />
<img width="325" alt="Screenshot 2025-05-20 at 13 25 50" src="https://github.com/user-attachments/assets/a1d04b5b-9977-4181-817a-6716893bbfee" />
<img width="265" alt="Screenshot 2025-05-20 at 13 25 57" src="https://github.com/user-attachments/assets/8079d00e-d068-4551-9961-49aec3ce6de4" />

chihsuan committed:
Change variable name

youknowriad commented on packages/dataviews/src/components/dataviews-filters/style.scss:
```diff
@@ -375,3 +376,11 @@
+	.components-input-control__prefix {
+		padding-left: $grid-unit-10;
+	}
+}
```
Needs an empty line

chihsuan replied:
Fixed in 53f987cfb452c62ef4f87f41816e5df25fa58b21.

youknowriad reviewed (approved):
Let's ship this and iterate.

chihsuan committed:
Fix lint

oandregal commented on packages/dataviews/src/components/dataviews-filters/filter.tsx:
```diff
@@ -27,6 +27,7 @@ const SPACE = ' ';
  * Internal dependencies
  */
 import SearchWidget from './search-widget';
+import InputWidget from './input-widget';
```
@chihsuan @youknowriad I want to make sure everyone is aware that every 2 week we sync changes from Gutenberg upstream into the `packages/dataviews` folder in calypso as per PCYsg-173c-p2.

Renaming a file is the kind of thing we shouldn't do. Most of the time it'll go fine, but it can create unnecessary work in these syncs — for example, if anyone creates a new file `filter-summary.tsx`, git won't be able to detect the original summary was renamed. My recommendation is that we revert this file rename.

I'm also not super stoked about other renames in this PR (`FilterSummary` to `Filter`, etc.) for the same reasons, but their surface for errors is minor — it's okay to keep them if you feel strongly about them. My general recommendation would be to make the minimal change that works and avoid needless refactors (renames, reorganizations, etc.).

This may sound like a little thing, but if we're not diligent with the kind of changes we make, we'll run into issues with the sync process.

oandregal commented:
@chihsuan @youknowriad this PR has made filters visible for any field that declares a type. This is problematic, as I [mentioned](https://github.com/Automattic/wp-calypso/pull/103276#discussion_r2092601407).  For example, this field declaration generates a filter and there's no way to opt-out:

```js
{
  id: 'date',
  label: 'Date',
  type: 'datetime',
}
```

I've started a draft to look into this at https://github.com/Automattic/wp-calypso/pull/103949

chihsuan commented:
> I've started a draft to look into this at https://github.com/Automattic/wp-calypso/pull/103949

Thanks for looking into this! 🙏  @oandregal 

>  the ability to have a filter that displays presets and the Edit function. I've been AFK and focused on other projects, so I may have missed some conversations: is anyone looking at this?

I will look at it, but I'm pausing for now because #103660 introduces "operators" to the Edit function. Before moving forward with the "date", I want to confirm that that is the correct approach.

As @youknowriad [commented](https://github.com/Automattic/wp-calypso/pull/103276#issuecomment-2886501430), for all types except "date", it should display either the "elements" in a select or listbox, or use a basic input field. For the "date" type, we can modify the Edit function to accept elements and render both the preset options and a calendar picker.

Additionally, we will need to render different controls for the "date" type based on the operator.
